File: python/remotetable/cell_types.py
"""Spreadsheet-ish type coerce (CONTRACT schema_version 1)."""

from __future__ import annotations

import re
from typing import List, Mapping, Optional, Sequence


def coerce(value: str, type_name: Optional[str]) -> str:
    t = (type_name or "").strip().lower()
    raw = value if value is not None else ""
    if t in ("", "string"):
        return raw
    if t == "number":
        return coerce_number(raw)
    if t == "timestamp":
        return coerce_timestamp(raw)
    if t == "checkbox":
        return coerce_checkbox(raw)
    return raw


def coerce_number(raw: str) -> str:
    s = (raw or "").strip()
    if not s:
        return ""
    cleaned = re.sub(r"[^\d.\-eE+]", "", s)
    if not cleaned:
        return s
    try:
        f = float(cleaned)
        if f == int(f) and abs(f) < 1e15:
            return str(int(f))
        return str(f)
    except ValueError:
        return s


def coerce_timestamp(raw: str) -> str:
    s = (raw or "").strip()
    if not s:
        return ""
    if s.isdigit():
        return s
    digits = "".join(c for c in s if c.isdigit())
    if len(digits) >= 13:
        return digits[:13]
    if len(digits) >= 10:
        return str(int(digits[:10]) * 1000)
    return s


def coerce_checkbox(raw: str) -> str:
    s = (raw or "").strip().lower()
    if not s:
        return "false"
    if s in ("true", "1", "yes", "y", "on", "checked"):
        return "true"
    if s in ("false", "0", "no", "n", "off", "unchecked"):
        return "false"
    return "true" if s in ("true", "1", "yes") else "false"


def coerce_row(
    headers: Sequence[str],
    row: Sequence[str],
    columns: Sequence[Mapping],
) -> List[str]:
    type_by = {}
    for c in columns or []:
        if isinstance(c, dict):
            type_by[c.get("name", "")] = c.get("type", "string")
    out = []
    for i, h in enumerate(headers):
        v = row[i] if i < len(row) else ""
        out.append(coerce(str(v), type_by.get(h)))
    return out

File: android/remotetable/src/main/java/com/davidelang/remotetable/CellTypes.kt
package com.davidelang.remotetable

/**
 * Spreadsheet-ish type coerce for wire cells (CONTRACT schema_version 1).
 * Unknown type → string (trimmed). Output is always a string cell value.
 */
object CellTypes {
    const val STRING = "string"
    const val NUMBER = "number"
    const val TIMESTAMP = "timestamp"
    const val CHECKBOX = "checkbox"

    fun coerce(value: String, type: String?): String {
        val t = type?.trim()?.lowercase().orEmpty()
        val raw = value
        return when (t) {
            "", STRING -> raw
            NUMBER -> coerceNumber(raw)
            TIMESTAMP -> coerceTimestamp(raw)
            CHECKBOX -> coerceCheckbox(raw)
            else -> raw
        }
    }

    fun coerceNumber(raw: String): String {
        val s = raw.trim()
        if (s.isEmpty()) return ""
        // strip currency symbols / spaces; keep digits, dot, minus
        val cleaned = s.replace(Regex("""[^\d.\-eE+]"""), "")
        if (cleaned.isEmpty()) return s
        return cleaned.toDoubleOrNull()?.let {
            // avoid scientific noise for whole numbers
            if (it == it.toLong().toDouble() && !it.isNaN() && it in Long.MIN_VALUE.toDouble()..Long.MAX_VALUE.toDouble()) {
                it.toLong().toString()
            } else {
                it.toString()
            }
        } ?: s
    }

    fun coerceTimestamp(raw: String): String {
        val s = raw.trim()
        if (s.isEmpty()) return ""
        s.toLongOrNull()?.let { return it.toString() }
        // ISO-8601-ish: extract digits for weak epoch ms if pure digits after strip
        val digits = s.filter { it.isDigit() }
        if (digits.length >= 10) {
            // prefer first 13 as millis, else 10 as seconds → millis
            return when {
                digits.length >= 13 -> digits.take(13)
                else -> (digits.take(10).toLongOrNull()?.times(1000L))?.toString() ?: s
            }
        }
        return s
    }

    fun coerceCheckbox(raw: String): String {
        val s = raw.trim()
        if (s.isEmpty()) return "false"
        val lower = s.lowercase()
        return when (lower) {
            "true", "1", "yes", "y", "on", "checked" -> "true"
            "false", "0", "no", "n", "off", "unchecked" -> "false"
            else -> if (RowOps.isTruthyTombstone(s, RowOps.DEFAULT_TRUE_VALUES)) "true" else "false"
        }
    }

    /** Coerce each cell in [row] by matching [headers] to [columns] (by dest column name). */
    fun coerceRow(
        headers: List<String>,
        row: List<String>,
        columns: List<ColumnDef>,
    ): List<String> {
        val typeByName = columns.associate { it.name to it.type }
        return headers.mapIndexed { i, h ->
            val v = row.getOrElse(i) { "" }
            coerce(v, typeByName[h])
        }
    }
}
